### ISSUE TITLE
How to deal with inherited structs?

### ISSUE THREAD

**Issue by fregate** [stephenberry/glaze#178]
Hi! 
I'm very exited about performance for encoding of this application and want to ask about how to use it in case like

```
class A
{
	int a;
	double b;

public:
	int a() const { return a; }
	double b() const { return b; }
};

template <>
struct glz::meta<A>
{
	static constexpr auto value = glz::object(
		"a", [](auto && v) { return v.a(); }, 
		"b", [](auto && v) { return v.b(); }
	);
};

class B : public A
{
	std::string c;

public:
	const std::string & c() const { return c; }
};

template <>
struct glz::meta<A>
{
	static constexpr auto value = glz::object("c", [](auto && v) { return v.c(); });
};

class Outer
{
	std::uinque_ptr<A> ptr;

public:
	Outer() { ptr = std::make_unique<B>(); }
	const std::uinque_ptr<const A> ptr() const { return ptr; }
}

template <>
struct glz::meta<Outer>
{
	static constexpr auto value = glz::object("ptr", [](auto && v) { return v.ptr(); }  //  ????
	);
};

int main(int, char **)
{
	Outer o;
	std::cout << glz::write_json(o);
	return 0;
}
```
What is the correct way to do this inheritance thing? In fact I have abstract class-interface and got errors like: 
```
write.hpp:646:63: error: cannot allocate an object of abstract type ‘MessagePayload’
common.hpp:749:30: error: cannot allocate an object of abstract type ‘MessagePayload’
```

I tried to output json string, but in result string I see string, not object (of course) `"payload":"{\"num\":1}"` - and this is not correct for me. 
In tests I can't find any related cases.

**Comment by fregate** [comment]
Make it variants? Like this `std::variant<var1_t, var2_t> v{};` 
I'll try

**Comment by fregate** [comment]
Ok. I've found good solution, make `glz::json_t json() override` in derived classes, and do something like 
```
return {
	{"num", number_}
};
```
Is there is way to get json_t from already defined `struct glz::meta<MyDerivedClass>`? And not build it from scratch in json function, like `return *this` or `return glz::meta<MyDerivedClass>(this)`?

**Comment by mwalcott3** [comment]
I'm going to chime in with my opinion on the matter but its really up to @stephenberry so Ill talk to him later about if and how we want to handle inheritance and polymorphism

Inheritance is not currently supported but wouldn't be that hard to handle. Polymorphism on the other hand is a massive pain since glaze relies heavily on knowing the type it is dealing with so It would need to be able to do visitation which would require knowing the types. Luckily we don't need to list all the types in one place like in a variant and we can leverage either a compile time or runtime type registry. I lean towards compile time but there are benefits to a runtime registry.

In my proposed implementation after the meta struct you would need to specify what classes it inherits from. If we want a compiletime registry there isn't really a good way of adding such a thing to glz::meta and the code to set it up is quite boilerplatey so unfortunately it would need to use macros.
In one possible implementation If B, C, and D all inherited from A you would need to put the following after each respective glz::meta.
```c++
//Something like this
//B meta
GLZ_INHERITANCE(B, A);
//C meta
GLZ_INHERITANCE(C, A);
//D meta
GLZ_INHERITANCE(D, A);
```
And for full support they would likely all need names added to the metas (static constexpr std::string_view name = "B";)

Alternatively we could add the parent to the meta and then just register the type:
```c++
//Something like this
template <>
struct glz::meta<B>
{
        using parent = A; // We could handle multiple inheritance by adding something like glz::type_list
        static constexpr std::string_view name = "B";
	static constexpr auto value = glz::object("c", [](auto && v) { return v.c(); });
};
GLZ_REGISTER(B);
```
The benefit of this is that we could add support for handling any registered object type in a std::any automatically and would be helpful if we wanted to add support for stuff like generating language bindings. This approach also wouldn't require registration unless polymorphism was used. I lean towards this approach. 

Another thing to point out in the code you posted is that since the lamdas in the glz::object are returning values they would only support writing out to json and not reading in. I think @stephenberry was planning on adding setter/getter support since it was the most general way of supporting bitfields but its not a pattern we normally use. I'm surprised noone has run into this yet.

> Is there is way to get json_t from already defined struct glz::meta<MyDerivedClass>? And not build it from scratch in json function, like return *this or return glz::meta<MyDerivedClass>(this)?
I'm sorry I don't quite understand.

**Comment by fregate** [comment]
> I'm sorry I don't quite understand.

I asked about - is there is way to obtain json_t (or smthing like that) from meta<MyStruct> and not build it in function from scratch?

**Comment by fregate** [comment]
Anyway, thank you for quick response!

**Comment by mwalcott3** [comment]
> I asked about - is there is way to obtain json_t (or smthing like that) from meta and not build it in function from scratch?
No unfortunately not. Ill keep it in mind though.

Ill try to add the polymorphism stuff in soon provided I get approval. Hopefully that will address your use case.

**Comment by mwalcott3** [comment]
We might also be able to handle this through having the user supply a read/write function in the base class that is called. The derived class can then call write/read with the correct type.

I didn't originally think of this because it doesn't allow deducing the type when reading and it would have to read to whatever type is currently stored but we cant entirely support changing the underlying type unless its being stored in a std::unique_ptr or std::sharered_ptr anyways.

**Comment by mwalcott3** [comment]
> I asked about - is there is way to obtain json_t (or smthing like that) from meta and not build it in function from scratch?

I dislike json_t but I think we should add a way of doing this since like you pointed out we have the structure information from the meta so as long as two things share a similar structure (And json_t should match any structure) we can generate the code to copy the data out. Something auto json = glz::convert<json_t>(*this);

We should probably add pmr support to address some of the performance concerns,

**Comment by mwalcott3** [comment]
I talked with @stephenberry and we don't want to rush into inheritance/polymorphism. We don't want to be adding major features right now. We will likely initially just go with your suggestion of creating a json_t from a meta type for now.

**Comment by fregate** [comment]
Ok. Thank you for response and new features to implement in future. I will try with `glz::raw_json` to understand how it works.
And of course, follow this project.

Anyway, I made some changes to compile it with gcc-10 (for my embedded-like project). Is it interesting or you targeting only for new compilers and standards?

**Comment by mwalcott3** [comment]
> Anyway, I made some changes to compile it with gcc-10 (for my embedded-like project). Is it interesting or you targeting only for new compilers and standards?
I don't think we were motivated enough to update glaze to support gcc10 but it sounds like you already did the heavy lifting. If you make a pull request or publish your fork on github (didnt see it) I'll review the changes and merge them provided everything looks good. I'll set up a ci build for gcc10 but I can't promise we won't drop support for it later if it's missing something we want to use.

**Comment by mwalcott3** [comment]
No macro solution to the type registry problem https://godbolt.org/z/onYTjhf66
